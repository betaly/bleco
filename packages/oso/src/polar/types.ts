export interface PolarResource {
  name: string;
  permissions?: string[];
  roles?: string[];
  rolePermissions?: Record<string, string[]>;
  roleInherits?: Record<string, string[]>;
  relations?: Record<string, {model: string; property: string}>;
}
