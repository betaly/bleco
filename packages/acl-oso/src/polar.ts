import {TextWriter} from '@yellicode/core';
import {ValueOrPromise} from '@loopback/context';
import {Policy} from '@bleco/acl';
import {StreamWriter} from '@obcode/templating';
import {BufferWriter} from 'memstreams';
import isEmpty from 'tily/is/empty';
import camelCaseName from 'tily/string/camelCaseName';

export type AnyPolicy = Omit<Policy, 'type'>;

export namespace PolarGeneration {
  export type GenerateCallback = (writer: TextWriter) => ValueOrPromise<void>;

  export async function generate(callback: GenerateCallback): Promise<string> {
    const stream = new BufferWriter();
    await callback(new StreamWriter(stream));
    return stream.data.toString();
  }

  export function generateActorScripts(writer: TextWriter, resource: AnyPolicy) {
    generateBlockScripts('actor', writer, resource);
  }

  export function generateResourceScripts(writer: TextWriter, resource: AnyPolicy) {
    generateBlockScripts('resource', writer, resource);
    generateHasRolesAndHasPermissions(writer, resource);
    generateCustomRules(writer, resource);
  }

  export function generateBlockScripts(type: string, writer: TextWriter, policy: AnyPolicy) {
    writer.writeLine(`${type} ${policy.model.name} {`);
    writer.increaseIndent();

    if (policy.roles) {
      writeInlineRoles(writer, policy.roles);
    }

    if (policy.permissions) {
      writeInlinePermissions(writer, policy.permissions);
    }

    if (policy.relations && !isEmpty(policy.relations)) {
      writeInlineRelations(writer, policy.relations, policy);
    }

    if (policy.rolePermissions) {
      writeInlineRules(writer, policy.rolePermissions);
    }

    if (policy.roleInherits) {
      writeInlineRules(writer, policy.roleInherits);
    }

    writer.decreaseIndent();
    writer.writeLine('}');

    // relations mapping
    if (policy.relations && !isEmpty(policy.relations)) {
      writeHasRelations(writer, policy.relations, policy);
    }
  }

  function generateCustomRules(writer: TextWriter, policy: AnyPolicy) {
    if (policy.rules?.oso) {
      policy.rules?.oso.forEach((rule: string) => writer.writeLine(rule));
    }
  }

  function generateHasRolesAndHasPermissions(writer: TextWriter, policy: AnyPolicy) {
    const res = camelCaseName(policy.model.name);
    const Res = policy.model.name;
    // reference: https://github.com/osohq/gitclub/blob/sam/custom-roles/backends/express-typeorm/src/authorization.polar
    const rules = `
has_role(actor: Actor, name: String, ${res}: ${Res}) if
  mapping in ${res}.principals and
  mapping matches {principalId: actor.id, roleId: name};
has_permission(actor: Actor, action: String, ${res}: ${Res}) if
  has_custom_role(actor, role, ${res}) and
  permission in role.permissions and
  permission.permission = action;
has_custom_role(actor: Actor, role: Role, ${res}: ${Res}) if
  role in ${res}.roles and
  mapping in role.principals and
  mapping.principalId = actor.id;`;
    writer.writeLine(rules);
  }

  function writeInlineRoles(writer: TextWriter, roles: string[]) {
    writer.writeLine(`roles = [${roles.map(r => `"${r}"`).join(', ')}];`);
  }

  function writeInlinePermissions(writer: TextWriter, permissions: string[]) {
    writer.writeLine(`permissions = [${permissions.map(p => `"${p}"`).join(', ')}];`);
  }

  function writeInlineRelations(writer: TextWriter, relations: string[], policy: AnyPolicy) {
    const definition = policy.model.definition;
    const content = relations.map(r => `${r}: ${definition.relations[r].target().name}`).join(', ');
    writer.writeLine(`relations = {${content}};`);
  }

  function writeInlineRules(writer: TextWriter, rules: Record<string, string[]>) {
    for (const [key, values] of Object.entries(rules)) {
      const [name, parent] = parseRoleName(key);
      for (const value of values) {
        let line = `"${value}" if "${name}"`;
        if (parent) {
          line += ` on "${parent}"`;
        }
        line += ';';
        writer.writeLine(line);
      }
    }
  }

  function writeHasRelations(writer: TextWriter, relations: string[], policy: AnyPolicy) {
    const model = policy.model;
    for (const name of relations) {
      const target = policy.model.definition.relations[name].target();
      const property = name;
      const targetVal = camelCaseName(target.name);
      writer.writeLine(
        `has_relation(${targetVal}: ${target.name}, "${name}", _: ${model.name}{${property}: ${targetVal}});`,
      );
    }
  }
}

function parseRoleName(roleName: string): [name: string, parent?: string] {
  if (roleName.includes('@')) {
    const parts = roleName.split('@');
    return [parts[0], parts[1]];
  } else if (roleName.includes('.')) {
    const parts = roleName.split('.');
    return [parts[1], parts[0]];
  }
  return [roleName];
}
