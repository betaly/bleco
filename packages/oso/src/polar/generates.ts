import {TextWriter} from '@obcode/templating';
import isEmpty from 'tily/is/empty';
import camelCaseName from 'tily/string/camelCaseName';
import {PolarResource} from './types';

export function generateResourceScripts(writer: TextWriter, resource: PolarResource) {
  generateResourceBlockScripts(writer, resource);
  generateHasRelationScripts(writer, resource);
}

export function generateResourceBlockScripts(writer: TextWriter, resource: PolarResource) {
  writer.writeLine(`resource ${resource.name} {`);
  writer.increaseIndent();

  if (resource.roles) {
    writer.writeLine(`roles = [${resource.roles.map(r => `"${r}"`).join(', ')}];`);
  }

  if (resource.permissions) {
    writer.writeLine(`permissions = [${resource.permissions.map(p => `"${p}"`).join(', ')}];`);
  }

  if (resource.relations && !isEmpty(resource.relations)) {
    writer.writeLine(`relations = {`);
    writer.increaseIndent();
    Object.keys(resource.relations).forEach((name, i, arr) => {
      const relation = resource.relations![name];
      let line = `${name}: ${relation.model}`;
      if (i < arr.length - 1) {
        line += ',';
      }
      writer.writeLine(line);
    });
    writer.decreaseIndent();
    writer.writeLine('};');
  }

  if (resource.rolePermissions) {
    for (const [role, permissions] of Object.entries(resource.rolePermissions)) {
      for (const permission of permissions) {
        writer.writeLine(`"${permission}" if "${role}";`);
      }
    }
  }

  if (resource.roleInherits) {
    for (const [role, inherits] of Object.entries(resource.roleInherits)) {
      const parts = role.split('@');
      const roleName = parts[0];
      const parentName = parts[1];
      for (const inherit of inherits) {
        let line = `"${inherit}" if "${roleName}"`;
        if (parentName) {
          line += ` on "${parentName}"`;
        }
        line += ';';
        writer.writeLine(line);
      }
    }
  }

  writer.decreaseIndent();
  writer.writeLine('}');
}

export function generateHasRelationScripts(writer: TextWriter, resource: PolarResource) {
  if (resource.relations && !isEmpty(resource.relations)) {
    for (const [name, relation] of Object.entries(resource.relations)) {
      const modelVar = camelCaseName(relation.model);
      writer.writeLine(
        `has_relation(${modelVar}: ${relation.model}, "${name}", _: ${resource.name}{${relation.property}: ${modelVar}});`,
      );
    }
  }
}
