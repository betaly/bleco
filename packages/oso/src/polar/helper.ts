import {TextWriter} from '@yellicode/core';
import {ValueOrPromise} from '@loopback/context';
import {BufferWriter} from 'memstreams';
import {StreamWriter} from '@obcode/templating';

export type GenerateScriptCallback = (writer: TextWriter) => ValueOrPromise<void>;

export async function generateScript(callback: GenerateScriptCallback): Promise<string> {
  const stream = new BufferWriter();
  await callback(new StreamWriter(stream));
  return stream.data.toString();
}
