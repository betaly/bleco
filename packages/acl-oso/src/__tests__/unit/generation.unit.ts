import {Repo} from '@bleco/acl/dist/test';
import {PolarGeneration} from '../../polar';

describe('generation', function () {
  describe('generate', function () {
    it('should generate with writer', async () => {
      const result = await PolarGeneration.generate(writer => {
        writer.write('Hello, ');
        writer.write('world!');
      });
      expect(result).toBe('Hello, world!');
    });
  });

  describe('generate polar script', function () {
    it('should generateResourceScripts', async () => {
      const result = await PolarGeneration.generate(writer => {
        PolarGeneration.generateResourceScripts(writer, {
          model: Repo,
          roles: ['admin', 'maintainer', 'reader'],
          relations: ['org'],
          permissions: [
            'read',
            'create_issues',
            'list_issues',
            'create_role_assignments',
            'list_role_assignments',
            'update_role_assignments',
            'delete_role_assignments',
          ],
          rolePermissions: {
            admin: [
              'create_role_assignments',
              'list_role_assignments',
              'update_role_assignments',
              'delete_role_assignments',
            ],
            reader: ['read', 'list_issues', 'create_issues'],
          },
          roleInherits: {
            'org.owner': ['admin'],
            'org.member': ['reader'],
            admin: ['maintainer'],
            maintainer: ['reader'],
          },
        });
      });
      // console.log(result);
      expect(result).toMatchSnapshot();
    });
  });
});
