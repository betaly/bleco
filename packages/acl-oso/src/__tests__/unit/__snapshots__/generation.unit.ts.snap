// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generation generate polar script should generateResourceScripts 1`] = `
"resource Repo {
	roles = [\\"admin\\", \\"maintainer\\", \\"reader\\"];
	permissions = [\\"read\\", \\"create_issues\\", \\"list_issues\\", \\"create_role_assignments\\", \\"list_role_assignments\\", \\"update_role_assignments\\", \\"delete_role_assignments\\"];
	relations = {org: Org};
	\\"create_role_assignments\\" if \\"admin\\";
	\\"list_role_assignments\\" if \\"admin\\";
	\\"update_role_assignments\\" if \\"admin\\";
	\\"delete_role_assignments\\" if \\"admin\\";
	\\"read\\" if \\"reader\\";
	\\"list_issues\\" if \\"reader\\";
	\\"create_issues\\" if \\"reader\\";
	\\"admin\\" if \\"owner\\" on \\"org\\";
	\\"reader\\" if \\"member\\" on \\"org\\";
	\\"maintainer\\" if \\"admin\\";
	\\"reader\\" if \\"maintainer\\";
}
has_relation(org: Org, \\"org\\", _: Repo{org: org});

has_role(actor: Actor, name: String, repo: Repo) if
  mapping in repo.principals and
  mapping matches {principalId: actor.id, roleId: name};
has_permission(actor: Actor, action: String, repo: Repo) if
  has_custom_role(actor, role, repo) and
  permission in role.permissions and
  permission.permission = action;
has_custom_role(actor: Actor, role: Role, repo: Repo) if
  role in repo.roles and
  mapping in role.principals and
  mapping.principalId = actor.id;
"
`;
